### PR TITLE
[KIECLOUD-678] - Create Seed Job for BAMOE images

### PR DESCRIPTION
**Pull request by spolti**
Signed-off-by: Your Name <fspolti@redhat.com>

**Thank you for submitting this pull request**

**JIRA**: _(please edit the JIRA link if it exists)_ 

[link](https://examle.com)

**referenced Pull Requests**: _(please edit the URLs of referenced pullrequests if they exist)_

* referenced pull request 1
* referenced pull request 2
* referenced pull request 2
etc. 

<pre>
How to retest a PR or trigger a specific build:

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>
</pre>

### CODE REVIEW

File: job-dsls/src/main/groovy/org/kie/jenkins/jobdsl/Constants.groovy [hunk 1]
```diff
@@ -51,6 +51,7 @@ class Constants {
     static final String KERBEROS_CRED = ''
     static final String IMAGE_REPO = ''
     static final String IMAGE_BRANCH = ''
+    static final String BAMOE_IMAGE_BRANCH = '7.13.x-blue'
```
Comment (davidesalerno) [inline job-dsls/src/main/groovy/org/kie/jenkins/jobdsl/Constants.groovy:54]: @Ginxo I've got a question: why don't we need to set the IMAGE_BRANCH constant that is used for the IMAGE_BRANCH string param on the RHPAM images? Where are we retrieving this param from tor the non BAMOE components?

Comment (spolti) [reply job-dsls/src/main/groovy/org/kie/jenkins/jobdsl/Constants.groovy:54]: it is retrieved in the download.sh using the PROD_VERSION.

Comment (davidesalerno) [reply job-dsls/src/main/groovy/org/kie/jenkins/jobdsl/Constants.groovy:54]: So we could add a BAMOE_IMAGE_BRANCH_EXT = '-blue' and append it to the IMAGE_BRANCH in order not to have to change it and have it automatically aligned, even if we won't need anything different from the current value.
Do you think that this change is worth?

Comment (spolti) [reply job-dsls/src/main/groovy/org/kie/jenkins/jobdsl/Constants.groovy:54]: it is not really needed, I've changed the download.sh script as well.
we can keep this new env or delete.

**Review by davidesalerno — APPROVED** [review]

File: job-dsls/jobs/osbs/bamoe_osbs_seed_cloud_jobs.groovy [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+                script('''
+
+                      private void validateParameters(required, optionals){
+                          // Check if all required params are supplied
+                          for ( param in required ) {
+                              def arg = param.value[0]
+                              def flag = param.value[1]
+                              if(!arg){
+                              error "$param.key parameter is required but was not specified."
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/bamoe_osbs_seed_cloud_jobs.groovy:64]:
```suggestion
                                  error "$param.key parameter is required but was not specified."
```

File: job-dsls/jobs/osbs/bamoe_osbs_seed_cloud_jobs.groovy [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+                              }
+                          }
+                      }
+
+                      private void mainProcess(user, password, keytab){
+
+                          """ Parameters available for the build. """
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/bamoe_osbs_seed_cloud_jobs.groovy:71]:
```suggestion
                          /// Parameters available for the build
```

File: job-dsls/jobs/osbs/bamoe_osbs_seed_cloud_jobs.groovy [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+                          def REQUIRED_DOWNLOAD_PARAMETERS = [
+                                  "PROD_VERSION": [PROD_VERSION, '-v'],
+                                  "PROD_COMPONENT": [PROD_COMPONENT, '-c'],
+                          ]
+
+                          def OPTIONAL_DOWNLOAD_PARAMETERS = [
+                                  "IMAGE_REPO": [IMAGE_REPO, '-r'],
+                                  "IMAGE_BRANCH": [IMAGE_BRANCH, '-n'],
+                                  "IMAGE_SUBDIR": [IMAGE_SUBDIR, '-d'],
+                          ]
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/bamoe_osbs_seed_cloud_jobs.groovy:103]:
```suggestion
                          def REQUIRED_BUILD_PARAMETERS = [
                                  'PROD_VERSION': [PROD_VERSION, '-v'],
                                  'PROD_COMPONENT': [PROD_COMPONENT, '-c'],
                                  'OSBS_BUILD_TARGET': [OSBS_BUILD_TARGET, '-t'],
                          ]

                          def OPTIONAL_BUILD_PARAMETERS = [
                                  'KERBEROS_PASSWORD': [password, '-s'],
                                  'KERBEROS_PRINCIPAL': [user, '-p'],
                                  'KERBEROS_KEYTAB': [keytab, '-k'],
                                  'OSBS_BUILD_USER': [OSBS_BUILD_USER, '-i'],
                                  'BUILD_DATE': [BUILD_DATE, '-b'],
                                  'GIT_USER': [GIT_USER, '-u'],
                                  'GIT_EMAIL': [GIT_EMAIL, '-e'],
                                  'CEKIT_BUILD_OPTIONS': [CEKIT_BUILD_OPTIONS, '-o'],
                                  'CEKIT_CACHE_LOCAL': [CEKIT_CACHE_LOCAL, '-l'],
                          ]

                          def OPTIONAL_BUILD_SWITCHES = [
                                  'VERBOSE': [VERBOSE, '-g'],
                          ]

                          def REQUIRED_DOWNLOAD_PARAMETERS = [
                                  'PROD_VERSION': [PROD_VERSION, '-v'],
                                  'PROD_COMPONENT': [PROD_COMPONENT, '-c'],
                          ]

                          def OPTIONAL_DOWNLOAD_PARAMETERS = [
                                  'IMAGE_REPO': [IMAGE_REPO, '-r'],
                                  'IMAGE_BRANCH': [IMAGE_BRANCH, '-n'],
                                  'IMAGE_SUBDIR': [IMAGE_SUBDIR, '-d'],
                          ]
```

File: job-dsls/jobs/osbs/bamoe_osbs_seed_cloud_jobs.groovy [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+                          // The download script is in the image, but build.sh and build-overrides.sh which it calls will be downloaded
+                          def download_command = "/opt/rhba/download.sh"
+                          def build_command = "build-osbs.sh"
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/bamoe_osbs_seed_cloud_jobs.groovy:107]:
```suggestion
                          def buildCommand = 'build-osbs.sh'
```
please replace the rest of `build_command` references

File: job-dsls/jobs/osbs/bamoe_osbs_seed_cloud_jobs.groovy [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+                      node("osbs-builder"){
+                          stage("building ${PROD_COMPONENT}") {
+                              ws {
+                                  // Add the working directory to the current path so any scripts dropped there are on the path
+                                  withEnv(["PATH+W=$WORKSPACE"]) {
+                                      sh 'rm -rf ${WORKSPACE}/{*,.*} || true'
+                                      if (env.KERBEROS_CRED) {
+                                          withCredentials([usernamePassword(credentialsId: env.KERBEROS_CRED, usernameVariable: 'USERNAME', passwordVariable: 'PASSWORD')]) {
+                                              mainProcess(USERNAME, PASSWORD, "")
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/bamoe_osbs_seed_cloud_jobs.groovy:172]:
```suggestion
                                              mainProcess(USERNAME, PASSWORD, '')
```

File: job-dsls/jobs/osbs/bamoe_osbs_seed_cloud_jobs.groovy [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+                                          withCredentials([file(credentialsId: env.KERBEROS_KEYTAB, variable: 'FILE')]) {
+                                              if (!env.KERBEROS_PRINCIPAL) {
+                                                    echo "Reading the Kerberos Principal from provided Keytab..."
+                                                    def get_principal_from_file = sh(returnStdout: true, script: \'\'\'
+                                                        #!/bin/bash
+                                                        klist -kt $FILE |grep REDHAT.COM | awk -F" " \'NR==1{print $4}\'
+                                                    \'\'\')
+                                                    env.KERBEROS_PRINCIPAL = get_principal_from_file.trim()
+                                              }  
+                                              mainProcess(env.KERBEROS_PRINCIPAL, "", FILE)
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/bamoe_osbs_seed_cloud_jobs.groovy:184]:
```suggestion
                                              mainProcess(env.KERBEROS_PRINCIPAL, '', FILE)
```

File: job-dsls/jobs/osbs/bamoe_osbs_seed_cloud_jobs.groovy [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+                                          }
+                                      } else {
+                                          error "Either KERBEROS_PRINCIPAL and KERBEROS_KEYTAB must be specified, or user/password with KERBEROS_CRED"
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/bamoe_osbs_seed_cloud_jobs.groovy:187]:
```suggestion
                                          error 'Either KERBEROS_PRINCIPAL and KERBEROS_KEYTAB must be specified, or user/password with KERBEROS_CRED'
```

File: job-dsls/jobs/osbs/bamoe_osbs_seed_cloud_jobs.groovy [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+            stringParam("OSBS_BUILD_USER", "${osbsBuildUser}")
+            stringParam("KERBEROS_KEYTAB", "${kerberosKeytab}")
+            stringParam("KERBEROS_CRED", "${kerberosCred}")
+            stringParam("IMAGE_REPO", "${imageRepo}")
+            stringParam("IMAGE_BRANCH", "${imageBranch}")
+            stringParam("IMAGE_SUBDIR", "${imageSubdir}")
+            stringParam("GIT_USER", "${gitUser}")
+            stringParam("GIT_EMAIL", "${gitEmail}")
+            stringParam("CEKIT_CACHE_LOCAL", "${cekitCacheLocal}")
+            stringParam("VERBOSE", "${verbose}")
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/bamoe_osbs_seed_cloud_jobs.groovy:50]:
```suggestion
            stringParam('BUILD_DATE', "${buildDate}")
            stringParam('PROD_VERSION', "${prodVersion}")
            stringParam('PROD_COMPONENT', "${Component}")
            stringParam('OSBS_BUILD_TARGET', "${osbsBuildTarget}")
            stringParam('CEKIT_BUILD_OPTIONS', "${cekitBuildOptions}")
            stringParam('KERBEROS_PRINCIPAL', "${kerberosPrincipal}")
            stringParam('OSBS_BUILD_USER', "${osbsBuildUser}")
            stringParam('KERBEROS_KEYTAB', "${kerberosKeytab}")
            stringParam('KERBEROS_CRED', "${kerberosCred}")
            stringParam('IMAGE_REPO', "${imageRepo}")
            stringParam('IMAGE_BRANCH', "${imageBranch}")
            stringParam('IMAGE_SUBDIR', "${imageSubdir}")
            stringParam('GIT_USER', "${gitUser}")
            stringParam('GIT_EMAIL', "${gitEmail}")
            stringParam('CEKIT_CACHE_LOCAL', "${cekitCacheLocal}")
            stringParam('VERBOSE', "${verbose}")
```

File: job-dsls/jobs/osbs/bamoe_osbs_seed_cloud_jobs.groovy [hunk 10]
```diff
@@ -0,0 +1,198 @@
+import org.kie.jenkins.jobdsl.Constants
+
+folder("OSBS")
+def folderPath = "OSBS"
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/bamoe_osbs_seed_cloud_jobs.groovy:4]:
```suggestion
def folderPath = 'OSBS'
folder(folderPath)
```

File: job-dsls/jobs/osbs/bamoe_osbs_seed_cloud_jobs.groovy [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+// Job Description
+String jobDescription = "Job responsible for seed jobs to building bamoe openshift image"
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/bamoe_osbs_seed_cloud_jobs.groovy:6]:
```suggestion
String jobDescription = 'Job responsible for seed jobs to building bamoe openshift image'
```

**Review by Ginxo — APPROVED** [review]
LGTM, just few minor technical suggestions

**Pull request merged**